src/Microsoft.CodeAnalysis.Remote.Razor/RazorLanguageService.cs

```diff
@@ -19,25 +21,69 @@ internal class RazorLanguageService : ServiceHubServiceBase
         public RazorLanguageService(Stream stream, IServiceProvider serviceProvider)
             : base(serviceProvider, stream)
         {
-            Rpc.JsonSerializer.Converters.Add(new RazorDiagnosticJsonConverter());
+            Rpc.JsonSerializer.Converters.Add(TagHelperDescriptorJsonConverter.Instance);
+            Rpc.JsonSerializer.Converters.Add(RazorDiagnosticJsonConverter.Instance);
+            Rpc.JsonSerializer.Converters.Add(RazorExtensionJsonConverter.Instance);
+            Rpc.JsonSerializer.Converters.Add(RazorConfigurationJsonConverter.Instance);
+            Rpc.JsonSerializer.Converters.Add(ProjectSnapshotJsonConverter.Instance);
 
             // Due to this issue - https://github.com/dotnet/roslyn/issues/16900#issuecomment-277378950
             // We need to manually start the RPC connection. Otherwise we'd be opting ourselves into 
             // race condition prone call paths.
             Rpc.StartListening();
+
+            Debugger.Launch();
         }
 
-        public async Task<TagHelperResolutionResult> GetTagHelpersAsync(Guid projectIdBytes, string projectDebugName, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<TagHelperResolutionResult> GetTagHelpersAsync(
+            Guid projectIdBytes, 
+            string projectDebugName, 
+            string factoryTypeName, 
+            ProjectSnapshot snapshot, 
+            CancellationToken cancellationToken = default(CancellationToken))
         {
             var projectId = ProjectId.CreateFromSerialized(projectIdBytes, projectDebugName);
 
             var solution = await GetSolutionAsync(cancellationToken).ConfigureAwait(false);
             var project = solution.GetProject(projectId);
 
-            var resolver = new DefaultTagHelperResolver(designTime: true);
-            var result = await resolver.GetTagHelpersAsync(project, cancellationToken).ConfigureAwait(false);
+            ((SerializedProjectSnapshot)snapshot).InitializeWorkspaceProject(project);
+
+            RazorTemplateEngine templateEngine = null;
+            if (factoryTypeName != null)
```

Comment:
We're being a bit sneaky. 

The VS side decides which factory to use to create the template engine, and then serializes the AQFN over to service hub. 

Then we rehydrate the factory on the service hub side. This relies on the assembly containing the factory to exist in the Load context, which we already do, and anyone shipping a VSIX can easily do as well.

Then we use the factory to configure the crap out of a new engine.

Then we find the tag helper providers in there and use them.

Reply:
We will do something a bit different when we want to ship this kind of extensibility in a nukpg. This is good enough for anyone shipping a VSIX to get started... ahem..